src/types/dict.rs

```diff
@@ -180,6 +180,15 @@ pub trait PyDictMethods<'py>: crate::sealed::Sealed {
     /// so long as the set of keys does not change.
     fn iter(&self) -> BoundDictIterator<'py>;
 
+    /// Iterates over the contents of this dictionary while holding a critical section on the dict.
+    /// This is useful when the GIL is disabled and the dictionary is shared between threads.
+    /// It is not guaranteed that the dictionary will not be modified during iteration when the
+    /// closure calls arbitrary Python code that releases the current critical section.
```

Comment:
New doc looks great 👍

I think it's potentially worth mentioning that this is roughly a performance optimization for stable where we can't optimise `.iter().try_for_each()` at the moment. (And that for infallible iteration, `.locked_for_each()` has no advantage over `.iter().for_each()`.)

Reply:
>  And that for infallible iteration, .locked_for_each() has no advantage over .iter().for_each().

I do not understand the second statement, could you elaborate why this is the case?

Reply:
I was thinking that we can optimize `.for_each` already on stable so if they don't need to deal with possible errors, then that's probably more optimal? Maybe I missed something.

Reply:
Ok I understand now, I will add it to the docs.